python/cudf/cudf/dataframe/dataframe.py

```diff
@@ -2158,27 +2163,23 @@ def join(
             cat_join = True
             lcats = lhs[idx_col_name].cat.categories
             rcats = rhs[idx_col_name].cat.categories
+
+            def set_categories(col, cats):
+                return col.cat._set_categories(cats, True).fillna(-1)
+
             if how == "left":
                 cats = lcats
-                rhs[idx_col_name] = (
-                    rhs[idx_col_name].cat._set_categories(cats).fillna(-1)
-                )
+                rhs[idx_col_name] = set_categories(rhs[idx_col_name], cats)
             elif how == "right":
                 cats = rcats
-                lhs[idx_col_name] = (
-                    lhs[idx_col_name].cat._set_categories(cats).fillna(-1)
-                )
+                lhs[idx_col_name] = set_categories(lhs[idx_col_name], cats)
             elif how in ["inner", "outer"]:
                 cats = sorted(set(lcats) | set(rcats))
```

Comment:
@kkraus14 this could probably be optimized now that `categories` is a Column

Reply:
This needs to be reworked now that categories is a Column. Calling `set()` will be a bad time.

Reply:
Though it seems we'd need to wait until https://github.com/rapidsai/cudf/pull/2068 is merged since we have some join tests that validate categoricals with different categorical dtypes.

Reply:
Given that the categories are no longer packed into the dtype you should be okay now, no?

Reply:
Oh my bad, I thought I saw a test that checked joining two dataframes with same-named categoricals of different dtypes. Seems we don't have that test.